GVFS/GVFS.Common/Tracing/QueuedPipeStringWriter.cs

```diff
@@ -0,0 +1,211 @@
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.IO.Pipes;
+using System.Text;
+using System.Threading;
+
+namespace GVFS.Common.Tracing
+{
+    public class QueuedPipeStringWriter : IDisposable
+    {
+        private static readonly TimeSpan PollingInterval = TimeSpan.FromSeconds(30);
+
+        private readonly Func<NamedPipeClientStream> pipeFactory;
+        private readonly ConcurrentQueue<string> queue = new ConcurrentQueue<string>();
+        private readonly AutoResetEvent queuePulse = new AutoResetEvent(initialState: false);
+
+        private Thread writerThread;
+        private NamedPipeClientStream pipeClient;
+        private bool isRunning;
+        private bool isDisposed;
+
+        public QueuedPipeStringWriter(Func<NamedPipeClientStream> pipeFactory)
+        {
+            this.pipeFactory = pipeFactory;
+        }
+
+        public event EventHandler<QueuedPipeStringWriterStateEventArgs> StateChanged;
+
+        public enum QueuedPipeStringWriterState
+        {
+            Stopped,
+            Failing,
+            Healthy,
+        }
+
+        public void Start()
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            if (this.isRunning)
+            {
+                return;
+            }
+
+            this.isRunning = true;
+
+            this.writerThread = new Thread(this.ThreadProc)
+            {
+                Name = nameof(QueuedPipeStringWriter),
+                IsBackground = true,
+            };
+
+            this.writerThread.Start();
+        }
+
+        public void Enqueue(string message)
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            if (!this.isRunning)
+            {
+                throw new InvalidOperationException($"{nameof(QueuedPipeStringWriter)} is not started");
+            }
+
+            this.queue.Enqueue(message);
+
+            // Wake the queue draining thread if it is waiting
+            this.queuePulse.Set();
+        }
+
+        public void Stop()
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            if (!this.isRunning)
+            {
+                return;
```

Comment:
Throw an exception here too? Or are there scenarios where we expect `Stop` to be called multiple times (or on a `QueuedPipeStringWriter` that was never started)?

Reply:
Same question as this [comment](https://github.com/Microsoft/VFSForGit/pull/935/files#r275741684).